File: JS/PassChecker/PasswordChecker.js
/**
 * Checks if a password meets certain criteria.
 * @param {string} password - The password to be checked.
 * @throws {Error} If the password is not a string, is too short, too long, or does not meet the required criteria.
 */
export function checkPassword(password) {
    if(typeof password !== 'string') {
        throw new Error('Password must be a string');
    }
    if(password.length < 8) {
        throw new Error('Password must be at least 8 characters long');
    }
    if(password.length > 20) {
        throw new Error('Password must be at most 20 characters long');
    }
    if(!/[a-z]/.test(password)) {
        throw new Error('Password must contain at least one lowercase letter');
    }
    if(!/[A-Z]/.test(password)) {
        throw new Error('Password must contain at least one uppercase letter');
    }
    if(!/[0-9]/.test(password)) {
        throw new Error('Password must contain at least one digit');
    }
    if(!/[^a-zA-Z0-9!]/.test(password)) {
        throw new Error('Password must contain at least one special character');
    }
}

/**
 * Checks if a password meets the specified criteria without using regular expressions.
 * 
 * @param {string} password - The password to be checked.
 * @throws {Error} If the password is not a string.
 * @throws {Error} If the password is less than 8 characters long.
 * @throws {Error} If the password is more than 20 characters long.
 * @throws {Error} If the password does not contain at least one lowercase letter.
 * @throws {Error} If the password does not contain at least one uppercase letter.
 * @throws {Error} If the password does not contain at least one digit.
 * @throws {Error} If the password does not contain at least one special character.
 */
export function checkPasswordWithoutRegEx(password) {
    if(typeof password !== 'string') {
        throw new Error('Password must be a string');
    }
    if(password.length < 8) {
        throw new Error('Password must be at least 8 characters long');
    }
    if(password.length > 20) {
        throw new Error('Password must be at most 20 characters long');
    }
    let hasLowercase = false;
    let hasUppercase = false;
    let hasDigit = false;
    let hasSpecialChar = false;
    
    for(let i = 0; i < password.length; i++) {
        const char = password[i];
        
        if(char >= 'a' && char <= 'z') {
            hasLowercase = true;
        } else if(char >= 'A' && char <= 'Z') {
            hasUppercase = true;
        } else if(char >= '0' && char <= '9') {
            hasDigit = true;
        } else {
            hasSpecialChar = true;
        }
    }
    
    if(!hasLowercase) {
        throw new Error('Password must contain at least one lowercase letter');
    }
    if(!hasUppercase) {
        throw new Error('Password must contain at least one uppercase letter');
    }
    if(!hasDigit) {
        throw new Error('Password must contain at least one digit');
    }
    if(!hasSpecialChar) {
        throw new Error('Password must contain at least one special character');
    }
}

File: Python/PassChecker/password_checker.py
import re

def validate_password(password):
    """
    Validates a password based on the following criteria:
    - Length must be between 8 and 20 characters
    - Must contain at least one digit
    - Must contain at least one uppercase letter
    - Must contain at least one lowercase letter
    - Must contain at least one special character from '!@#$%^&*()_+-=[]{}|:;"\'<>,.?/~`'
    
    Args:
        password (str): The password to be validated.
    
    Returns:
        bool: True if the password is valid, False otherwise.
    """
    if len(password) < 8 or len(password) > 20:
        return False
    
    if not any(char.isdigit() for char in password):
        return False
    
    if not any(char.isupper() for char in password):
        return False
    
    if not any(char.islower() for char in password):
        return False
    
    if not any(char in '!@#$%^&*()_+-=[]{}|:;"\'<>,.?/~`' for char in password):
        return False
    
    return True


def validate_password_with_error_messages(password):
    """
    Validates a password based on the following criteria:
    - Length must be between 8 and 20 characters
    - Must contain at least one digit
    - Must contain at least one uppercase letter
    - Must contain at least one lowercase letter
    - Must contain at least one special character from '!@#$%^&*()_+-=[]{}|:;"\'<>,.?/`'
    
    Args:
        password (str): The password to be validated.
    
    Raises:
        ValueError: If the password does not meet any of the above criteria.
    """
    if len(password) < 8 or len(password) > 20:
        raise ValueError("Password length must be between 8 and 20 characters")
    
    if not any(char.isdigit() for char in password):
        raise ValueError("Password must contain at least one digit")
    
    if not any(char.isupper() for char in password):
        raise ValueError("Password must contain at least one uppercase letter")
    
    if not any(char.islower() for char in password):
        raise ValueError("Password must contain at least one lowercase letter")
    
    special_chars = '!@#$%^&*()_+-=[]{}|:;"\'<>,.?/`'
    if not any(char in special_chars for char in password):
        raise ValueError(f"Password must contain at least one special character from {special_chars}")
    
def validate_password_with_error_messages_regex(password):
    """
    Validates a password using regular expressions and raises ValueError with specific error messages for each validation rule.

    Args:
        password (str): The password to be validated.

    Raises:
        ValueError: If any of the validation rules fail.

    """
    if not re.search(r".{8,20}", password):
        raise ValueError("Password length must be between 8 and 20 characters")
    
    if not re.search(r"\d", password):
        raise ValueError("Password must contain at least one digit")
    
    if not re.search(r"[A-Z]", password):
        raise ValueError("Password must contain at least one uppercase letter")
    
    if not re.search(r"[a-z]", password):
        raise ValueError("Password must contain at least one lowercase letter")
    
    special_chars = r"!@#$%^&*()_+-=[]{}|:;\"'<>,.?/~`"
    if not re.search(f"[{re.escape(special_chars)}]", password):
        raise ValueError(f"Password must contain at least one special character from {special_chars}")
    
    
